### PR TITLE
Add axios retry interceptor

### LINKED ISSUE #265: Add configurable retry logic to mirror node requests
**Is your feature request related to a problem? Please describe.**
The mirror node can occasionally return non-2xx response codes due to transient errors like 502s for when proxy attempts to route to a pod that's in the middle of terminating.

**Describe the solution you'd like**
Add an axios retry interceptor that will retry failures that aren't a result of user input. It should:
* Retry 429 (rate limit) and any 5xx
* Default to trying 3 times (total) with exponential backoff in between
* Support customizing the above defaults

**Describe alternatives you've considered**
Retry in the application

**Additional context**

### PR DESCRIPTION
**Pull request by georgi-l95**
Signed-off-by: georgi-l95 <glazarov95@gmail.com>

**Description**:
Add an axios retry interceptor that will retry failures that aren't a result of user input. Default values for retries, which can be customised. Exponential back-off

**Related issue(s)**:

Fixes #265 

**Notes for reviewer**:
Extended some of the tests to fit this new functionality. Others which test for bad requests cover this already.

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: packages/providers/src.ts/base-provider.ts [hunk 1]
```diff
@@ -186,9 +195,22 @@ export interface Avatar {
 }
 
 export interface ProviderOptions {
-    headers?: Record<string, string>
+    headers: Record<string, string>
+    retry: RetryOptions
 }
 
+export interface RetryOptions {
```
Comment (natanasow) [inline packages/providers/src.ts/base-provider.ts:202]: Do we need to export it? As far as I see we are not using it in another package, so exporting is not mandatory.

File: packages/providers/src.ts/base-provider.ts [hunk 2]
```diff
@@ -186,9 +195,22 @@ export interface Avatar {
...
+    maxAttempts: number,
+    waitTime: number,
+    errorCodes: Array<number>
+}
+
+export const DEFAULT_RETRY_OPTIONS = {
```
Comment (natanasow) [inline packages/providers/src.ts/base-provider.ts:208]: Same as above.

File: packages/providers/src.ts/base-provider.ts [hunk 3]
```diff
@@ -218,28 +240,29 @@ export class BaseProvider extends Provider {
...
-
-        this._options = options || {};
         this._events = [];
         this._emittedEvents = {};
         this._previousPollingTimestamps = {};
         this.formatter = new.target.getFormatter();
         // If network is any, this Provider allows the underlying
         // network to change dynamically, and we auto-detect the
         // current network
+        if(options !== undefined && options !== null){
```
Comment (natanasow) [inline packages/providers/src.ts/base-provider.ts:257]: Can be simplified like that :point_down: 
```
this._options = { headers : {}, retry: DEFAULT_RETRY_OPTIONS };
if (typeof options === 'object' && Object.keys(options).length) {
  this._options = options;
}
```
or even better
```
this._options = (typeof options === 'object' && Object.keys(options).length) 
    ? options 
    : { headers : {}, retry: DEFAULT_RETRY_OPTIONS };
```

**Review by natanasow — APPROVED** [review]
LG, some nits

File: packages/providers/src.ts/base-provider.ts [hunk 4]
```diff
@@ -278,8 +301,47 @@ export class BaseProvider extends Provider {
...
+            return response;
+        }, error => {
+            const { config} = error;
+            const originalRequest = config;
+            if (error.response) {
+                if (this._options.retry.errorCodes.includes(error.response.status) && originalRequest._retry) {
+                    originalRequest._retriedRequest = true;
+                    originalRequest._attempts += 1;
+                    originalRequest._waitTime = originalRequest._waitTime * originalRequest._attempts;
+                    if(originalRequest._attempts === this._options.retry.maxAttempts){
```
Comment (ar-conmit) [inline packages/providers/src.ts/base-provider.ts:329]: nit: the comparison operator can be changed to `>=` to be more strictly

**Review by ar-conmit — APPROVED** [review]
LGTM

File: packages/providers/src.ts/base-provider.ts [hunk 5]
```diff
@@ -218,28 +240,29 @@ export class BaseProvider extends Provider {
...
-        this._options = options || {};
         this._events = [];
         this._emittedEvents = {};
         this._previousPollingTimestamps = {};
         this.formatter = new.target.getFormatter();
         // If network is any, this Provider allows the underlying
         // network to change dynamically, and we auto-detect the
         // current network
+        if(options !== undefined && options !== null){
+            this._options = Object.keys(options).length !== 0 ? options : { headers : {}, retry: DEFAULT_RETRY_OPTIONS };
```
Comment (Ivo-Yankov) [inline packages/providers/src.ts/base-provider.ts:258]: Q: Would this work if `options` is passed with only some custom `headers` specified, without a `retry` property?

Comment (georgi-l95) [reply packages/providers/src.ts/base-provider.ts]: No, but it's fixed now. If you don't specify `retry` property, it will take the default.

**Review by natanasow — APPROVED** [review]

**Pull request merged**